### DISCUSSION TITLE
task input needed for UI Stage of Engine layers and others

### DISCUSSION THREAD

**Discussion by jmscraig** [MicroTrendsLtd/NinjaTrader8#50] [General]
While I am here.  

Just thought i would ask if there any lingering topics you are still puzzling with and would like some input on?

Have a great day!

James

**Comment by MicroTrendsTom** [comment]
> 
in fact i do....  did you ever do something crazy such as check in a project delete all local source only to find out you didnt add a file such as a WPF user control in fact a chart ATM controller....  any ideas on reverse engineering it to decompile out the xaml etc?

if so this may lead to a bigger project layer for this project -  
plus the orders later
The chart UI for control in realtime etc
then a signals/filters/exits layer
all the jazz from here #31 

from there it will allow a developer to create strategies in 1 or 2 lines of code via deriving from the framework and filling in the trade entry signal rule...

so if you are up for that i have an idea whereby we can perhaps refactor a system built atop of this engine, chopout a lot of things simplify it in to layers  and then open source it all here... now that would be great and getting the darn chart trader control back would be a key to that etc

perhaps that would need to be in a private project to start with then when i can chop it down to the layers and cut out the redundant stuff for a 3rd party project it can be brought over to here for public digest that would be a real win win

and would lead to what i used to provide an algo trading framework for rapid code and test with hybrid realtime features etc
like a NT8 software dev kit for advanced unmanaged strategies that can be backtested and used in live realtime

i guess the route forwards is to provide the installed end product and access to the code base in TFS and go from there if you are interested.

note: the commercial engines i created are all realtime only and so there is no conflict.
i will look for ways to release those as open source foundations later on - but for now the above path is easier and builds atop of the current

**Reply by MicroTrendsTom** [reply]
This component set on the right... is one of many UI designs from MicroTrends used by whitelabels and vendors and so on..
but this is very basic and would need some evolving...   
![image](https://user-images.githubusercontent.com/24366913/153252242-90e51b0a-9eb8-42b8-8bbc-cf8d6fa028bd.png)
so what i would do is if we can get that source or replicate it again i would distill out the layers we need to do all this #31 

it also has an easy config for Realtime settings fast load and saving etc - this one was used on a customer project but the code components are my property and rights so any clutter or junk would be stripped out and made generic. 

![image](https://user-images.githubusercontent.com/24366913/153253087-e70bc0fe-a7cd-4b2c-8308-eaa82a50c1bb.png)

anyway if you able to assist pls let me know and we can start it off and open it all up
i would need your microsoft account details and add you with MTL dev team access and go from there etc

**Reply by jmscraig** [reply]
"any ideas on reverse engineering it to decompile out the xaml etc?"   Well I have not heard of this but will google around

**Reply by jmscraig** [reply]
It may not be easy to find a Strong solution to reverse engineer to build the WPF framework.   Especially given that NT customized the WPF offering they delivered so it can not be managed by the WPF controls tool vendor they use or any other package so I believe it would be a custom job to write a parser and do all the mapping.  Might be a lot easier to write a parser to get the majore outline and then fill the detail from there.  

Do you have a working version and it is just the source that is gone?

Thus far this is the strongest inputs I have found for a build it yourself approach.   

1) https://actualrandy.wordpress.com/2014/10/01/auto-generate-forms-and-class-by-reverse-engineering-your-class-or-web-reference/


2) https://stackoverflow.com/posts/14587563/timeline

In WPF you can dynamically create controls using XAML. Simply create a XAML snippet with the layout you want (either programatically or by hand) and use [XamlReader.Parse](http://msdn.microsoft.com/en-us/library/cc663033.aspx) to create the entire tree from the snippet. The returned object can then be inserted somewhere in the visual tree of your window.

To generate events from the visual tree generated by XamlReader you can use [routed events](http://wpftutorial.net/RoutedEvents.html).

**Reply by MicroTrendsTom** [reply]
- yes the working version i have it was above in the images etc
ok all good apparently i got the tools to do it reverse out baml to xaml  its called ILSpy  - will have to  have ok 
but if not ok i build it again from scratch and probably use DevExpress for some of it i do use that as my core 3rd party toolset
Yes i would definetly try to avoid NT8 WPF as much as possible...  external gui you can dock it alongside the chart window externally all good or ok inline in a new colum in the WPF  "grid"

Some more UIs
This is for a prop scalper hybrid approach -  this was really bespoke high end 
https://microtrends.co/wp-content/uploads/2016/11/automated-trading-systems-development-1.png

basic versions of a framework i used to whitelable for vendors such as BWT 
https://microtrends.co/wp-content/uploads/2017/01/NinjaTrader7-BWT-Precision-AutoTrader-7.png

**Reply by jmscraig** [reply]
"Some more UIs"  Very nice.  People see screens like that and say "Wow, that is far too complex, the screen needs to be more simple."   In reality it takes that level of detail and more for robust algorithmic trading.

**Reply by MicroTrendsTom** [reply]
yeah its like ms word approach  -give the user all they could possibly want  - then simplify views - basic and advanced - track events usage and have basi views and advanced etc  -  some people want notepad ok thats the purest basic vanila.  
evolve it organically etc - traders dont know what they want you have to invent on their behalf etc.
then in time they got exactly what they need and can use it in live trading effectively etc.

So i have opened a discussion here and posted a few case examples #52 


So for this discussion to recap:
For NT8 project powered by this NT8 strategy Engine - i wish to create a UI layer and make i available to trader/devs here.
See https://github.com/MicroTrendsLtd/NinjaTrader8/discussions/50#discussioncomment-2143069

1. a WPF NT8  UI for Trade and Config was created and code lost -  binaries are here in [dropbox UI](https://www.dropbox.com/s/3dhhsz1fuptxotj/BWT.NT8.GUI.UserControls.dll?dl=0) 
2. Believe i have found a BAML TO XAML exporter as used by MS itself in 2020/2021 pres release as an addin forVS - [ILSpy](https://github.com/icsharpcode/ILSpy)  
3.  Hopefully wont have to start from scratch on it... time will tell for anyone whos advanced and wants to come as a guest MicroTrends developer we can look at source code which calls it and explore with hot load in VS2019 etc so i can give access to TFS with a MS Account.... PM me if you want etc..

**Comment by MicroTrendsTom** [comment]
So for this discussion to recap:
For NT8 project powered by this NT8 strategy Engine - i wish to create a UI layer and make i available to trader/devs here.
See https://github.com/MicroTrendsLtd/NinjaTrader8/discussions/50#discussioncomment-2143069

a WPF NT8 UI for Trade and Config was created and code lost - binaries are here in [dropbox UI](https://www.dropbox.com/s/3dhhsz1fuptxotj/BWT.NT8.GUI.UserControls.dll?dl=0)
Believe i have found a BAML TO XAML exporter as used by MS itself in 2020/2021 pres release as an addin forVS - [ILSpy](https://github.com/icsharpcode/ILSpy)
Hopefully wont have to start from scratch on it... time will tell for anyone whos advanced and wants to come as a guest MicroTrends developer we can look at source code which calls it and explore with hot load in VS2019 etc so i can give access to TFS with a MS Account.... PM me if you want etc..

**Reply by jmscraig** [reply]
Hi Tom, 

I put some time into this.

I acquired the proper dependent WPF DLL and produce decompiles with no errors from ILSpy and also the most recent dotPeek app.  I am using Visual Studio 2022.


![TM_WPF-Deccomplile](https://user-images.githubusercontent.com/15313578/154334110-a22dc612-d274-4d80-a4b2-930e21c04be4.jpg)



I tried migration from .NET Framework 4.5 to 4.8 and saw the same complication errors reported below.

I tried migration to C#8,  and then C#7.3, these produced additional code breaking changes so dropped back to C#5 & C#6.

After a few hours of work I am still unable identify the source of a few build errors.    While there were no reported errors during final decompiles
 it feels like the code is not quite right.

The ILSpy decompile to C#5 and C#6 o NET 4.5 & 4.8 builds clean when the two XAML files are excluded from the project.  With the default inclusion of the XAML files the errors start with CS0260  where the compiler is seeing an second instance of the class and so is reporting that the only true instance should include a 'partial' modifier in the class constructor.

"error CS0260: Missing partial modifier on declaration of type 'UserControlEasyConfig'; another partial declaration of this type exists"
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs0260

Each of the projects in attached zips where built in short directors right off the root of C:, for example C:\MT-WPF-C.6"  

Here are the ILSpy no-error decompiles to NET4.5 with C#5 and C#6
[MT_WPF-C#6.zip](https://github.com/MicroTrendsLtd/NinjaTrader8/files/8082531/MT_WPF-C.6.zip)
[MT-WPF-C#5.zip](https://github.com/MicroTrendsLtd/NinjaTrader8/files/8082532/MT-WPF-C.5.zip)


The dotPeek decompile error list does not include CS0260 but starts with similar complaints about duplicate definition of the interface "IComponentConnector.Connect" and duplicate definitions of the WPF form action objects like "grdMain".  

"error CS8646: 'IComponentConnector.Connect(int, object)' is explicitly implemented more than once."
error CS0102: The type 'UserControlEasyConfig' already contains a definition for 'grdMain'
https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0102
 
Here is the no-error decompile from dotPeek.
[TM-WPF-C#5-dotPeek.zip](https://github.com/MicroTrendsLtd/NinjaTrader8/files/8082590/TM-WPF-C.5-dotPeek.zip)


I need to switch to address a few topics so passing these over to see if your eyes and knowledge of the code base render the  remaining issues quickly and easily identifiable.  

If I read your original note correctly you still have the original *.cs files and just needed the XAML.  Hopefully at least one of the attached three sets of XAML files provides some value there.   If you have similar XAML then a diff between those and these might surface decompile errors.   Also, it would be interesting to see how well, how accurately ILSpy & dotPeek decompiled the attached *.cs files vs the originals.

James

**Reply by MicroTrendsTom** [reply]
Ah ok wow!!!!!! great work !!!!! speechless, mind blown that's amazing.... :-)


EDITED>>>>
Yes i found half of it in TFS seems i did a 50% mistake... 

I have the solution and project:
with these parts:
UserControlNinjaBuddy.xaml
UserControlNinjaBuddy.cs

was only Missing these:
userControlEasyConfig.xaml
userControlEasyConfig.cs

investigating and testing now   and rebuilding based on XAML as the code behind is not coming through - but its not a biggie i can decompile the CS also and read the il more or less to work out what i did and write it again :-)
>>>>

Ok lets do this, as a the chief collaborator in the dev dept....  certainly you are now a VIP in my book and if it appeals you dont need to pay for your "DIY lease" or software from MicroTrends or ATS,   - AFT, AST, AWTweb etc - 
let's have you onboard as  "Algo Dev -"  in  the Dev  role  or something in the trading group...  so you get it all free and  betas and maybe even join in the coding in parts as you are here at least etc     So just drop me a line PM and we can sort that out etc 

really big thanks for the collaboration  you rock!

Anyway you are giving me enthusiasm! to get on this again amidst the craziness elsewhere :-) 
I will also wheel back to this and read it and go through it properly I'm mid release atm

**Comment by MicroTrendsTom** [comment]
Update wow... what a headache this stuff is lol!  NT WPF and the property grid but will perserver
In the reverse compiled out Duplicates - Yes so i can see the duplicates are due to the VS generating code and properties again - eg autogenerated code etc    UserControlEasyConfig.g.i.cs

So i have regenerated  -recoded it but cannot compile it due to some crazy situation with the propertygrid.... hahaha
PropertyGrid not found in namespace yet intellisense shows it just fine...
Ok so it might just have to go into the mix as loose XAML and get loaded at runtime  instead of a sideby side control dll etc
e.g Uri resourceLocator = new Uri("/..UserControls;component/usercontroleasyconfig.xaml", UriKind.Relative);
Application.LoadComponent(this, resourceLocator); and so on...  so that means putting a void IComponentConnector.Connect(int connectionId, object target) method in... should be ok

**Comment by MicroTrendsTom** [comment]
ok checked in a compliable - what should be a clone type deal -  in the project 
https://github.com/MicroTrendsLtd/NinjaTrader8/tree/main/ATSQuadroStrategyBase.GUI

so the idea from here is the output the ATSQuadroStrategyBase.GUI.dll would be used in NinjaTrader custom as  a side by side assembly and so thusly needs to be added as a reference within NT8 Code editor...  
\Documents\NinjaTrader 8\bin\Custom\ATSQuadroStrategyBase.GUI.dll 

but....if folks want to make it more open so its compiled and loaded at runtime that can be done also i guess easy enough just add in the partial autogen code from the "initialiaze" and idk but that should work with some tinkering i guess.

**Comment by MicroTrendsTom** [comment]
Folks thanks to the collaboration of @jmscraig this gave me the inertia to see this through to a point where it will be plain sailing from hereon.  Sample needs polishing  but is workable, for a test. you get the latest checked in code  you will see 2 folders added to "Strategies"
![image](https://user-images.githubusercontent.com/24366913/154786912-6b0097ed-37e1-4326-a5bf-4cd47e01926f.png)

Test the following sample to see:  ATSSamplePriceReversalTestHybrid
 - this was derived  from the class "ATSQuadroStrategyBaseTradingUILayerBase.cs" which provided Hybrid Algo  UI  features for semi auto, realtime interactive etc.  
![image](https://user-images.githubusercontent.com/24366913/154786853-fe37519c-8219-4e03-9d01-a5445a7fa622.png)

 The Box Breakout is not plumbed in or  "trail trg" button
 
![image](https://user-images.githubusercontent.com/24366913/154786957-2300c132-eff3-4be5-8625-c166010e166e.png)

Realtime Settings - Config  - basic to start with a multi tab and views of properties or a vertical grid would be better -over time it can be - assistance welcome as that is a real brain drain to get something good and easy to work with.
![image](https://user-images.githubusercontent.com/24366913/154786970-2d19801e-ec92-4651-a134-a2bb221c7baf.png)
- tread Carefully what you change when running as it will crash out the Strategy

**Reply by jmscraig** [reply]
Hi Tom, 

I have been playing the TestHybrid via Market Replay.   

This is fantastic.

James
